projects/ngx-skeleton-loader/src/lib/ngx-skeleton-loader.component.spec.ts

```diff
@@ -190,3 +195,25 @@ describe('NgxSkeletonLoaderComponent', () => {
     });
   });
 });
+
+describe('NgxSkeletonLoaderComponent2', () => {
```

Comment:
Actually, I looked for a way to put the content of this describe inside the other one and overwrite the testing module in the one test that I set up for the provided configuration but I couldn't make it work. Any idea? @willmendesneto @dev054

Reply:
In that case, you can have that added inside the `describe`, but adding some description (BDD like) to have the intentions more explicit, like we are doing here, in the assertion above in this file https://github.com/willmendesneto/ngx-skeleton-loader/pull/77/files/b2c51b63049a0b79c301ab605f499da83c056855#diff-6d738b6b6abf8c7c5e01ffd9918916062e651449f843b0d2f2d4e10dc5939d96R187-R196

```suggestion
describe('When the component receives a different default via module configuration', () => {
  beforeEach(
    waitForAsync(() => {
      fixture = TestBed.configureTestingModule({
        declarations: [ContainerComponent, NgxSkeletonLoaderComponent],
        providers: [
          { provide: PLATFORM_ID, useValue: 'browser' },
          { provide: NGX_SKELETON_LOADER_CONFIG, useValue: { appearance: 'circle', count: 3 }}
        ],
      }).createComponent(ContainerComponent);
      fixture.detectChanges();
    }),
  );

  it('should render skeleton with the provided config', () => {
    expect(fixture.nativeElement.querySelectorAll('.skeletons-with-provided-config .loader.circle').length).toBe(3);
  });
});
```

Reply:
I have tried that already before pushing my commit on the component's tests. it seems like it is not possible to change the TestBed's testing module after it has been created in the first beforeEach.

Reply:
I thought about `TestBed.overrideProvider` but we have another problem with it: the tests are not properly covered.
```
Chrome 91.0.4472.101 (Windows 10) NgxSkeletonLoaderComponent When the component receives a different default via module configuration should render skeleton with the provided config FAILED
        Error: Expected 0 to be 3.
            at <Jasmine>
            at UserContext.<anonymous> (http://localhost:9876/_karma_webpack_/src/lib/ngx-skeleton-loader.component.spec.ts:207:111)
            at ZoneDelegate.invoke (http://localhost:9876/_karma_webpack_/C:/git/ngx-skeleton-loader/node_modules/zone.js/dist/zone-evergreen.js:372:1)
            at ProxyZoneSpec.push.../../node_modules/zone.js/dist/zone-testing.js.ProxyZoneSpec.onInvoke (http://localhost:9876/_karma_webpack_/C:/git/ngx-skeleton-loader/node_modules/zone.js/dist/zone-testing.js:301:1)
Chrome 91.0.4472.101 (Windows 10): Executed 12 of 22 (1 FAILED) (0 secs / 0.718 secs)
Chrome 91.0.4472.101 (Windows 10) NgxSkeletonLoaderComponent When the component receives a different default via module configuration should render skeleton with the provided config FAILED
        Error: Expected 0 to be 3.
            at <Jasmine>
            at UserContext.<anonymous> (http://localhost:9876/_karma_webpack_/src/lib/ngx-skeleton-loader.component.spec.ts:207:111)
            at ZoneDelegate.invoke (http://localhost:9876/_karma_webpack_/C:/git/ngx-skeleton-loader/node_modules/zone.js/dist/zone-evergreen.js:372:1)
            at ProxyZoneSpec.push.../../node_modules/zone.js/dist/zone-testing.js.ProxyZoneSpec.onInvoke (http://localhost:9876/_karma_webpack_/C:/git/ngx-skeleton-loader/node_moduChrome 91.0.4472.101 (Windows 10): Executed 22 of 22 (1 FAILED) (1.184 secs / 1.009 secs)
TOTAL: 1 FAILED, 21 SUCCESS
TOTAL: 1 FAILED, 21 SUCCESS
TOTAL: 1 FAILED, 21 SUCCESS
npm ERR! Test failed.  See above for more details.
```

Reply:
I can see the issue you're facing and the reasons why. The idea is not creating a new `describe`, but reusing the one that exists.

So the tests will be updated by

- Removing the second `described` added
- Reusing the existent `describe` and passing forRoot config
- Adding the specific assertion on the existent describe

Let me play with this branch a little, so I'll push the changes as a commit into this pull request to get a the example included here, if you don't mind

Reply:
I'd be happy to read the code resolving the issue from you! Feel free to push anything new to this branch :smile:

Reply:
Just got some time to check that now, but you were faster than me :) Resolving this conversation, thanks for that!

Reply:
I just added some split between scenarios. So now we can have the checks that are component's common behaviours + no config and another describing checking how it should work when using `forRoot`

```ts
describe('NgxSkeletonLoaderComponent', () => {
  // tslint:disable-next-line: no-any
  let fixture: any;
  beforeEach(() => {
    spyOn(console, 'error');
  });

  describe('When the component uses default configuration', () => {
    beforeEach(
      waitForAsync(() => {
        fixture = TestBed.configureTestingModule({
          declarations: [ContainerComponent, NgxSkeletonLoaderComponent],
          providers: [{ provide: PLATFORM_ID, useValue: 'browser' }],
        }).createComponent(ContainerComponent);
        fixture.detectChanges();
      }),
    );

    it('should console 3 errors if `animation`, `appearance` and `count` receives invalid options and is running in development mode', () => {
      expect(console.error).toHaveBeenCalledTimes(3);
    });

    it('should console errors if `animation` is an invalid option and is running in development mode', () => {
      expect(console.error).toHaveBeenCalledWith(
        // tslint:disable-next-line: max-line-length
        `\`NgxSkeletonLoaderComponent\` need to receive 'animation' as: progress, progress-dark, pulse, false. Forcing default to "progress".`,
      );
    });

    it('should console errors if `count` is an invalid option and is running in development mode', () => {
      expect(console.error).toHaveBeenCalledWith(
        // tslint:disable-next-line: max-line-length
        `\`NgxSkeletonLoaderComponent\` need to receive 'count' a numeric value. Forcing default to "1".`,
      );
    });

    it('should console errors if `appearance` is an invalid option and is running in development mode', () => {
      expect(console.error).toHaveBeenCalledWith(
        // tslint:disable-next-line: max-line-length
        `\`NgxSkeletonLoaderComponent\` need to receive 'appearance' as: circle or line or empty string. Forcing default to "''".`,
      );
    });

    it('should add all relevant WAI-ARIA `aria-` attributes in all ngx-skeleton-loader', () => {
      expect(fixture.nativeElement.querySelectorAll('[aria-busy="true"]').length).toBe(15);
      expect(fixture.nativeElement.querySelectorAll('[aria-valuemin="0"]').length).toBe(15);
      expect(fixture.nativeElement.querySelectorAll('[aria-valuemax="100"]').length).toBe(15);
      expect(fixture.nativeElement.querySelectorAll('[aria-valuetext]').length).toBe(15);
      expect(fixture.nativeElement.querySelectorAll('[role="progressbar"]').length).toBe(15);
      expect(fixture.nativeElement.querySelectorAll('[tabindex="0"]').length).toBe(15);
    });

    it('should use progress as default animation if `animation` is not passed as component attribute', () => {
      expect(fixture.nativeElement.querySelectorAll('.skeletons-defaults .loader.progress').length).toBe(1);
    });

    describe('When skeleton is created using default settings', () => {
      it('should render a single skeleton', () => {
        expect(fixture.nativeElement.querySelectorAll('.skeletons-defaults .loader').length).toBe(1);
      });
    });

    describe('When skeleton is created passing loading text to be used as WAI-ARIA `aria-valuetext`', () => {
      it('should render a single skeleton', () => {
        expect(fixture.nativeElement.querySelectorAll('[aria-valuetext="Loading. Please wait ..."]').length).toBe(1);
      });
    });

    describe('When skeleton is created with count', () => {
      it('should render skeleton based on given count attribute', () => {
        expect(fixture.nativeElement.querySelectorAll('.skeletons-with-count .loader').length).toBe(2);
      });
    });

    describe('When skeleton is created with circle appearance', () => {
      it('should add styles based on circle class on the skeleton components', () => {
        expect(fixture.nativeElement.querySelectorAll('.skeletons-appearance-circle .loader.circle').length).toBe(1);
      });
    });

    describe('When skeleton is created without animation', () => {
      it('should NOT add progress animation styles based on animation class on the skeleton components', () => {
        expect(
          fixture.nativeElement.querySelectorAll('.skeletons-animation-no-animation .loader:not(.animation)').length,
        ).toBe(1);
      });

      it('should NOT add progress animation styles based on animation class if animation value is passed via binding', () => {
        expect(
          fixture.nativeElement.querySelectorAll(
            '.skeletons-animation-no-animation-via-binding .loader:not(.animation)',
          ).length,
        ).toBe(1);
      });
    });

    describe('When skeleton is created using `pulse` as animation', () => {
      it('should add pulse animation styles based on animation class on the skeleton components', () => {
        expect(fixture.nativeElement.querySelectorAll('.skeletons-animation-pulse .loader.pulse').length).toBe(1);
      });
    });

    describe('When skeleton is created using `progress-dark` as animation', () => {
      it('should add progress-dark animation styles based on animation class on the skeleton components', () => {
        expect(
          fixture.nativeElement.querySelectorAll('.skeletons-animation-progress-dark .loader.progress-dark').length,
        ).toBe(1);
      });
    });

    describe('When skeleton is created using `progress` as animation', () => {
      it('should add progress animation styles based on animation class on the skeleton components', () => {
        expect(fixture.nativeElement.querySelectorAll('.skeletons-animation-progress .loader.progress').length).toBe(1);
      });
    });

    describe('When skeleton is created with theming', () => {
      it('should render skeleton with styles based on theme attribute', () => {
        const skeletonWithTheming = fixture.nativeElement.querySelector('.skeletons-with-theming .loader.circle')
          .attributes as NamedNodeMap;

        expect((skeletonWithTheming.getNamedItem('style') as Attr).value).toBe(
          'width: 70px; height: 70px; border-radius: 10px;',
        );
      });
    });
  });

  describe('When the component receives a different default via module configuration', () => {
    beforeEach(
      waitForAsync(() => {
        fixture = TestBed.configureTestingModule({
          declarations: [ContainerComponent, NgxSkeletonLoaderComponent],
          providers: [
            { provide: PLATFORM_ID, useValue: 'browser' },
            { provide: NGX_SKELETON_LOADER_CONFIG, useValue: { appearance: 'circle', count: 3 } },
          ],
        }).createComponent(ContainerComponent);
        fixture.detectChanges();
      }),
    );

    it('should render skeleton with the provided config', () => {
      expect(fixture.nativeElement.querySelectorAll('.skeletons-with-provided-config .loader.circle').length).toBe(3);
    });
  });
});

```